src/sqlfluff/core/linter/fix.py

```diff
@@ -0,0 +1,308 @@
+        anchor_id = fix.anchor.uuid
+        anchor_info[anchor_id].add(fix)
+    return dict(anchor_info)
+
+
+def apply_fixes(
+    segment: BaseSegment,
+    dialect: "Dialect",
+    rule_code: str,
+    fixes: Dict[int, AnchorEditInfo],
+) -> Tuple["BaseSegment", List["BaseSegment"], List["BaseSegment"], bool]:
+    """Apply a dictionary of fixes to this segment.
+
+    Used in applying fixes if we're fixing linting errors.
+    If anything changes, this should return a new version of the segment
+    rather than mutating the original.
+
+    Note: We need to have fixes to apply AND this must have children. In the case
+    of raw segments, they will be replaced or removed by their parent and
+    so this function should just return self.
+    """
+    if not fixes or segment.is_raw():
+        return segment, [], [], True
+
+    seg_buffer = []
+    before = []
+    after = []
+    fixes_applied: List[LintFix] = []
+    requires_validate = False
+
+    for seg in segment.segments:
+        # Look for uuid match.
+        # This handles potential positioning ambiguity.
+        anchor_info: Optional[AnchorEditInfo] = fixes.pop(seg.uuid, None)
+
+        if anchor_info is None:
+            # No fix matches here, just add the segment and move on.
+            seg_buffer.append(seg)
+            continue
+
+        # Otherwise there is a fix match.
+        seg_fixes = anchor_info.fixes
+        if (
+            len(seg_fixes) == 2 and seg_fixes[0].edit_type == "create_after"
+        ):  # pragma: no cover
+            # Must be create_before & create_after. Swap so the
+            # "before" comes first.
+            seg_fixes.reverse()
+
+        for f in anchor_info.fixes:
+            assert f.anchor.uuid == seg.uuid
+            fixes_applied.append(f)
+            linter_logger.debug(
+                "Matched fix for %s against segment: %s -> %s",
+                rule_code,
+                f,
+                seg,
+            )
+
+            # Deletes are easy.
+            if f.edit_type == "delete":
+                # We're just getting rid of this segment.
+                requires_validate = True
+                # NOTE: We don't add the segment in this case.
+                continue
+
+            # Otherwise it must be a replace or a create.
+            assert f.edit_type in (
+                "replace",
+                "create_before",
+                "create_after",
+            ), f"Unexpected edit_type: {f.edit_type!r} in {f!r}"
+
+            if f.edit_type == "create_after" and len(anchor_info.fixes) == 1:
+                # in the case of a creation after that is not part
+                # of a create_before/create_after pair, also add
+                # this segment before the edit.
+                seg_buffer.append(seg)
+
+            # We're doing a replacement (it could be a single
+            # segment or an iterable)
+            assert f.edit, f"Edit {f.edit_type!r} requires `edit`."
```

Comment:
Nice if the type system makes it impossible to get in a bad state

Reply:
This one might be tricky 🤔